app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkTouchHelper.kt

```diff
@@ -0,0 +1,111 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.library.bookmarks
+
+import android.graphics.Canvas
+import android.graphics.drawable.Drawable
+import androidx.appcompat.content.res.AppCompatResources
+import androidx.recyclerview.widget.ItemTouchHelper
+import androidx.recyclerview.widget.RecyclerView
+import mozilla.components.support.ktx.android.content.getColorFromAttr
+import mozilla.components.support.ktx.android.content.getDrawableWithTint
+import mozilla.components.support.ktx.android.util.dpToPx
+import org.mozilla.fenix.R
+import org.mozilla.fenix.home.sessioncontrol.SwipeToDeleteCallback
+import org.mozilla.fenix.library.bookmarks.viewholders.BookmarkNodeViewHolder
+import org.mozilla.fenix.library.bookmarks.viewholders.BookmarkSeparatorViewHolder
+
+class BookmarkTouchHelper(interactor: BookmarkViewInteractor) :
+    ItemTouchHelper(BookmarkTouchCallback(interactor))
+
+class BookmarkTouchCallback(private val interactor: BookmarkViewInteractor) :
+    ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
+
+    override fun getSwipeDirs(
+        recyclerView: RecyclerView,
+        viewHolder: RecyclerView.ViewHolder
+    ): Int {
+        val item = (viewHolder as BookmarkNodeViewHolder).item
+        return if (viewHolder is BookmarkSeparatorViewHolder || item?.inRoots() == true) {
+            0
+        } else {
+            super.getSwipeDirs(recyclerView, viewHolder)
+        }
+    }
+
+    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
+        val item = (viewHolder as BookmarkNodeViewHolder).item
+        item?.let { interactor.onDelete(setOf(it)) }
+    }
+
+    override fun onChildDraw(
+        c: Canvas,
+        recyclerView: RecyclerView,
+        viewHolder: RecyclerView.ViewHolder,
+        dX: Float,
+        dY: Float,
+        actionState: Int,
+        isCurrentlyActive: Boolean
+    ) {
+        super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
+        val icon = recyclerView.context.getDrawableWithTint(
+            R.drawable.ic_delete,
+            recyclerView.context.getColorFromAttr(R.attr.destructive)
+        )!!
+        val background = AppCompatResources.getDrawable(
+            recyclerView.context,
+            R.drawable.swipe_delete_background
+        )!!
+        val itemView = viewHolder.itemView
+        val iconLeft: Int
+        val iconRight: Int
+        val margin =
+            SwipeToDeleteCallback.MARGIN.dpToPx(recyclerView.context.resources.displayMetrics)
+        val iconWidth = icon.intrinsicWidth
+        val iconHeight = icon.intrinsicHeight
+        val cellHeight = itemView.bottom - itemView.top
+        val iconTop = itemView.top + (cellHeight - iconHeight) / 2
+        val iconBottom = iconTop + iconHeight
```

Comment:
I think we can separate some things out here to make it more readable.
* I like that you've made `val`s for these so that the `when` statement is more readable
* Some things that don't depend on the _immediate_ updated state of the draw function, such as `margin`, can be moved to private `val`s outside of the function
* Keep the more complicated `val`s like `iconTop`
* The simpler onces, like `itemView`, can just be done inline

Reply:
Tried to keep variables that do common calculations like `iconTop` and `iconBottom` here and move the rest inside the `when`. Can't remove `margin` because it needs a Context.